File: v2/clients/ui/src/lib/catalog.ts
/**
 * Reading a registry's catalog.json — the marketplace WITHOUT a daemon.
 *
 * The daemon reaches the same rows over `marketplace.catalog` (it builds them from the index it
 * verified). This module is the other door: a plain fetch of the static file the publish service
 * writes beside index.json, for the public marketplace page, which has no socket, no session and
 * no keys.
 *
 * IT VERIFIES NOTHING, deliberately. The page renders links; it never downloads or executes a
 * bundle. Signature checking belongs where execution is — the daemon on install, the installer
 * stub on download — and a browser checking a signature it then ignores is theatre. Nothing here
 * may ever become the input to an install.
 */

import type { CatalogBundle } from '../gateway/protocol'

/** The generated document. Mirrors agent_runtime/domain/catalog.py — change both together. */
export interface CatalogDoc {
  schema: number
  registry: string
  publisher: string
  /** what relative artifact urls resolve against ('' => the catalog's own location) */
  base: string
  /** where web-delivered agents run ('' => no Open-in-browser links anywhere) */
  webHost: string
  bundles: CatalogBundle[]
}

/** The shape number this client understands. A newer document is refused rather than guessed at. */
export const SUPPORTED_SCHEMA = 1

const EMPTY: CatalogDoc = { schema: SUPPORTED_SCHEMA, registry: '', publisher: '', base: '', webHost: '', bundles: [] }

/**
 * Join one artifact url against the document's base.
 *
 * Rows are relative when the registry is a directory (it has to keep working after being copied)
 * and absolute when the writer knew the public base. `new URL` handles both: an absolute url
 * ignores the base entirely. The fallback is the catalog's own address, which is the right answer
 * for a registry served out of its own folder.
 */
export function assetUrl(doc: CatalogDoc, url: string, catalogUrl = ''): string {
  if (!url) return ''
  try {
    return new URL(url, doc.base || catalogUrl || location.href).toString()
  } catch {
    return url
  }
}

/**
 * Fetch and validate a catalog.
 *
 * Throws with a message meant for a reader, not a console: this is the ONE request the public
 * page makes, so its failure is the whole page's failure and "Failed to fetch" tells a visitor
 * nothing about whether the store is empty, moved, or broken.
 */
export async function fetchCatalog(url: string): Promise<CatalogDoc> {
  let response: Response
  try {
    response = await fetch(url, { cache: 'no-cache' })
  } catch {
    throw new Error(`Could not reach the marketplace registry at ${url}.`)
  }
  if (!response.ok) {
    // A 404 is the common one and it is not an error the visitor caused: it is a registry that
    // has never been published to.
    if (response.status === 404) return { ...EMPTY }
    throw new Error(`The marketplace registry answered ${response.status} for ${url}.`)
  }
  let doc: CatalogDoc
  try {
    doc = (await response.json()) as CatalogDoc
  } catch {
    throw new Error('The marketplace registry returned something that is not a catalog.')
  }
  if (!Array.isArray(doc?.bundles)) {
    throw new Error('The marketplace registry returned a catalog with no bundles list.')
  }
  if (Number(doc.schema || 1) > SUPPORTED_SCHEMA) {
    throw new Error(
      `This page reads catalog schema ${SUPPORTED_SCHEMA} and the registry published ${doc.schema}. ` +
        'Reload — the page is probably a cached older build.'
    )
  }
  return { ...EMPTY, ...doc }
}

File: v2/agent_runtime/domain/catalog.py
"""The STORE VIEW of a registry index — one finished row per bundle, ready to render.

WHY THIS IS ITS OWN MODULE. A store card needs three things an index entry does not carry
directly: the creator's NAME (the entry names them by opaque id; only the signed roster has the
name), an installer URL a browser can follow (the entry's is relative to the index), and the
Open-in-browser link (joined from the index-level web host and the bundle id). Working those out
is a join over the whole index, not a property of one entry — so it lived in the daemon's
marketplace service, which meant the ONLY way to see a store was to run a daemon.

It is pure and it lives in the domain, so exactly ONE implementation of "what a store row is"
feeds all three consumers:

  * the DAEMON, on every `marketplace.catalog` call, over an index it has verified;
  * the PUBLISH SERVICE, which writes the result to ``catalog.json`` beside ``index.json``;
  * the PUBLIC MARKETPLACE PAGE, which fetches that file and has no daemon and no keys.

WHAT THIS IS NOT. It is not a trust boundary. Nothing here verifies anything — verification is
the registry client's job and happens where a bundle is DOWNLOADED and RUN. That is why the
daemon builds its rows from the index it just verified rather than reading ``catalog.json``: the
generated file is a rendering convenience for a page that only ever draws links, and a client
that installs software must never take its facts from it.

RELATIVE OR ABSOLUTE URLS. Index urls are relative so the same folder works from disk and from a
CDN, and joining them belongs to whoever knows where the index came from — the registry client
for the daemon, the index store for the publish service. Both hand that knowledge in here; a
directory build passes neither and its rows stay relative, which is correct for a folder that
still has to work after being uploaded somewhere else. ``base`` is recorded in the document so a
reader that receives relative rows knows what to join them against.
"""

from __future__ import annotations

from collections.abc import Callable
from urllib.parse import urljoin

from agent_runtime.domain.bundle import RegistryEntry, RegistryIndex

# The name of the generated file, beside index.json. Named here because both the writer
# (publish) and every reader (the web page, the uploader) must spell it the same way.
CATALOG_FILENAME = "catalog.json"

# Bumped only when a row's SHAPE changes incompatibly. A reader that does not recognise the
# number should fall back to index.json rather than render a store it may be misreading.
CATALOG_SCHEMA = 1


def build_catalog(
    index: RegistryIndex,
    *,
    base: str = "",
    resolve: Callable[[str], str] | None = None,
) -> dict:
    """The whole store, as plain JSON-able data.

    :param base: the url relative artifact paths resolve against ("" => leave them relative).
    :param resolve: an explicit joiner, when the caller knows better than ``base`` can express
        (the daemon's registry client, which knows the exact index url it fetched).
    """
    join = resolve or (lambda url: urljoin(base, url) if base and url else url)
    names = index.publishers.display_names() if index.publishers else {}
    return {
        "schema": CATALOG_SCHEMA,
        "registry": index.name,
        "publisher": index.publisher,
        # Recorded rather than applied when there is nothing to apply: a reader holding relative
        # rows and an empty base joins against the catalog's own location, which is the right
        # answer for a registry served straight out of its own directory.
        "base": base,
        "webHost": index.web_host,
        "bundles": [_row(entry, join, index, names) for entry in index.bundles],
    }


def _row(
    entry: RegistryEntry,
    join: Callable[[str], str],
    index: RegistryIndex,
    publisher_names: dict[str, str],
) -> dict:
    """One catalog row."""
    out = {
        "id": entry.id,
        "name": entry.name,
        "version": entry.version,
        "description": entry.description,
        "agentdCompat": entry.agentd_compat,
        "price": entry.price,
        "entitlement": entry.entitlement,
        "size": entry.size,
        "icon": entry.icon,
        "delivery": {"web": entry.delivery.web, "exe": entry.delivery.exe},
        # WHO STANDS BEHIND IT. `publisherId` is the opaque creator id the entry's signature was
        # verified against — the identity that is actually proven — and `publisher` is the name to
        # put in front of a reader. Three sources, in falling order of how much they are worth:
        #
        #   1. the SIGNED roster, looked up by creator id — the only verified name there is;
        #   2. the entry's own declared name, which an operator-built index writes from bundle.toml
        #      and which therefore comes from the same party that built the index;
        #   3. the index's single publisher, the schema-1 answer to a question schema 1 asks once.
        #
        # An entry that names a creator NEVER falls past step 1. A missing roster name there means
        # the roster does not list them, and the honest render for that is the bare id — dropping
        # to a name they typed themselves would dress an unlisted creator as a known one.
        "publisherId": entry.publisher_id,
        "publisher": publisher_names.get(entry.publisher_id, "")
        or ("" if entry.publisher_id else (entry.publisher or index.publisher)),
    }
    if entry.delivery.web and index.web_host:
        out["webUrl"] = f"{index.web_host.rstrip('/')}/apps/{entry.id}/"
    if entry.installers:
        out["installers"] = [
            {
                "platform": asset.platform,
                "url": join(asset.url),
                "size": asset.size,
                "sha256": asset.sha256,
            }
            for asset in entry.installers
        ]
    return out
